File: hoffman-agents-node/src/core/trace-buffer.js
const crypto = require('crypto');

class TraceEvent {
  constructor(fromState, toState, timestamp, prediction, predictionCorrect, predictionError, token) {
    this.fromState = fromState;
    this.toState = toState;
    this.timestamp = timestamp;
    this.prediction = prediction;
    this.predictionCorrect = predictionCorrect;
    this.predictionError = predictionError;
    this.token = token ?? null;
    if (predictionError < 0 || predictionError > 1) {
      throw new Error(`prediction_error must be in [0, 1], got ${predictionError}`);
    }
  }
}

class TraceBuffer {
  constructor(maxlen = 50) {
    if (maxlen < 1) throw new Error(`maxlen must be >= 1, got ${maxlen}`);
    this._maxlen = maxlen;
    this._events = [];
    this._cursor = 0;
  }

  append(event) {
    if (this._events.length < this._maxlen) {
      this._events.push(event);
    } else {
      this._events[this._cursor] = event;
      this._cursor = (this._cursor + 1) % this._maxlen;
    }
  }

  getRecent(n) {
    if (n <= 0) return [];
    n = Math.min(n, this._events.length);
    if (this._events.length < this._maxlen) {
      return this._events.slice(-n);
    }
    const ordered = this._events.slice(this._cursor).concat(this._events.slice(0, this._cursor));
    return ordered.slice(-n);
  }

  asStateSequence() {
    return this._events.map(e => e.toState);
  }

  predictionErrorMean(window = 20) {
    const recent = this.getRecent(window);
    if (recent.length === 0) return 0;
    return recent.reduce((s, e) => s + e.predictionError, 0) / recent.length;
  }

  isFull() { return this._events.length >= this._maxlen; }

  get length() { return this._events.length; }
  get maxlen() { return this._maxlen; }

  *[Symbol.iterator]() {
    if (this._events.length < this._maxlen) {
      yield* this._events;
    } else {
      yield* this._events.slice(this._cursor);
      yield* this._events.slice(0, this._cursor);
    }
  }

  clear() {
    this._events = [];
    this._cursor = 0;
  }

  static fromEvents(events, maxlen) {
    const buf = new TraceBuffer(maxlen ?? events.length);
    for (const e of events) buf.append(e);
    return buf;
  }
}

module.exports = { TraceEvent, TraceBuffer };

File: hoffman-agents-python/src/conscious_agent/core/trace_buffer.py
from __future__ import annotations

from collections.abc import Iterator
from dataclasses import dataclass, field
from typing import Self


@dataclass
class TraceEvent:
    from_state: int
    to_state: int
    timestamp: int
    prediction: int
    prediction_correct: bool
    prediction_error: float
    token: str | None = None

    def __post_init__(self) -> None:
        if not 0.0 <= self.prediction_error <= 1.0:
            raise ValueError(f"prediction_error must be in [0, 1], got {self.prediction_error}")


class TraceBuffer:
    def __init__(self, maxlen: int = 50) -> None:
        if maxlen < 1:
            raise ValueError(f"maxlen must be >= 1, got {maxlen}")
        self._maxlen = maxlen
        self._events: list[TraceEvent] = []
        self._cursor = 0

    def append(self, event: TraceEvent) -> None:
        if len(self._events) < self._maxlen:
            self._events.append(event)
        else:
            self._events[self._cursor] = event
            self._cursor = (self._cursor + 1) % self._maxlen

    def get_recent(self, n: int) -> list[TraceEvent]:
        if n <= 0:
            return []
        n = min(n, len(self._events))
        if len(self._events) < self._maxlen:
            return self._events[-n:]
        return (self._events[self._cursor:] + self._events[:self._cursor])[-n:]

    def as_state_sequence(self) -> list[int]:
        return [e.to_state for e in self]

    def prediction_error_mean(self, window: int = 20) -> float:
        recent = self.get_recent(window)
        if not recent:
            return 0.0
        return sum(e.prediction_error for e in recent) / len(recent)

    def is_full(self) -> bool:
        return len(self._events) >= self._maxlen

    def __len__(self) -> int:
        return len(self._events)

    def __iter__(self) -> Iterator[TraceEvent]:
        if len(self._events) < self._maxlen:
            yield from self._events
        else:
            yield from self._events[self._cursor:]
            yield from self._events[:self._cursor]

    def __getitem__(self, index: int) -> TraceEvent:
        if not (0 <= index < len(self._events)):
            raise IndexError("TraceBuffer index out of range")
        if len(self._events) < self._maxlen:
            return self._events[index]
        return self._events[(self._cursor + index) % self._maxlen]

    @property
    def maxlen(self) -> int:
        return self._maxlen

    def resize(self, new_maxlen: int) -> None:
        if new_maxlen < 1:
            raise ValueError(f"maxlen must be >= 1, got {new_maxlen}")
        ordered = list(self)
        self._maxlen = new_maxlen
        if len(ordered) > new_maxlen:
            self._events = ordered[-new_maxlen:]
        else:
            self._events = ordered
        self._cursor = 0

    def clear(self) -> None:
        self._events.clear()
        self._cursor = 0

    @classmethod
    def from_events(cls, events: list[TraceEvent], maxlen: int | None = None) -> Self:
        if maxlen is None:
            maxlen = len(events)
        buf = cls(maxlen=maxlen)
        for e in events:
            buf.append(e)
        return buf
